File: 019_remove_nth_node_from_end_of_the_list/two_ptrs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        """
        Complexity: O(n) one pass.
        Space: O(1) two ptrs.
        """
        # two pointers
        # ptr start from n element after the origal pos
        # new ptr start from n element and would point to the element before the removed element.
        ptr = head
        new_ptr = head
        # move ptr to be n element after the new_head
        # 1->2->3....-> n+1 (ptr) 
        # remove 2 if n+1 is the end
        for i in range(n):
            ptr = ptr.next
            if ptr is None: # if we need to remove the first one
                return head.next
            
        while head is not None:
            if ptr.next is not None:
                # ptr: 1->2->3...->n+1
                # new_ptr: 1 
                # if ptr.next, new_ptr: 1->2, 
                new_ptr = new_ptr.next
                ptr = ptr.next  
            else: 
                # if ptr: 1->2->3...->last element
                # new ptr: 1->2->3...->element before the removed one, new_ptr->next is on the removed one
                new_ptr.next = new_ptr.next.next 
                break

        return head

File: 019_remove_nth_node_from_end_of_the_list/two_ptrs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/*
 * Complexity: O(n) the length of the list
 * Space: O(1)
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* start = new ListNode(0);
        start -> next = head;
        ListNode* slow = start;
        ListNode* fast = start;
        // move the fast pointer to the nth pos from the head.
        for(int i = 1; i <= n; ++i){
            fast = fast -> next;
        }
        // the distance between fast and slow is n. thus, when fast reach the end, the slow would be in the l-n pos
        // then slow -> next would be the element that need to be removed
        while(fast -> next != NULL){
            slow = slow -> next;
            fast = fast -> next;
        }
        // removed the first element
        if(slow == start) return head -> next;
        slow -> next = slow -> next -> next;
        return head;     
    }
};
